src/ra_utility.h

```diff
@@ -172,6 +177,146 @@ _NODISCARD inline auto __cdecl tstrlen(_In_z_ LPCTSTR _Str) noexcept
+///   Up to <c><paramref name="bufsz" /> - 1</c> characters may be written, plus the null terminator.
+///   If the size is <c>0</c> or greater than <c>RSIZE_MAX</c>, the operation is undefined.
+/// </param>
+/// <param name="format">
+///   Pointer to a null-terminated multibyte string or wide-character string specifying how to
+///   interpret the data. The format string consists of ordinary multibyte characters(except %),
+///   which are copied unchanged into the output stream, and conversion specifications. The %n
+///   specifier is considered undefined.
+/// </param>
+/// <param name="...args">
+///   Forwarded arguments specifying data to print. If the number of arguments is more than the
+///   number of format specifiers present, the operation is undefined.
+/// </param>
+/// <returns>
+///   The number of characters written to <paramref name="buffer" /> excluding the null character.
+/// </returns>
+/// <exception cref="std::runtime_error">The function has an encoding error.</exception>
+/// <exception cref="std::invalid_argument">
+///   <para>
+///     Thrown <paramref name="buffer" />, <paramref name="format" />, or any argument in
+///     <paramref name="...args" /> are <c>nullptr</c>.
+///   </para>
+/// </exception>
+/// <exception cref="std::out_of_range">
+///   <para><paramref name="bufsz" /> does not meet the specification.</para>
+///   <para>
+///     The string stored in <paramref name="buffer" /> is larger than <paramref name="bufsz" />.
+///   </para>
+/// </exception>
+/// <exception cref="std::logic_error">
+///   Number arguments from <paramref name="...args" /> is greater than format specifiers present.
+/// </exception>
+/// <remarks>Exceptions are only explicitly thrown in release mode.</remarks>
+template<typename CharT = TCHAR, typename ...Args, typename = std::enable_if_t<is_char_v<CharT>>>
+_Success_(return >= 0) /*_NODISCARD*/ inline auto __cdecl
+stprintf_s(_Inout_updates_bytes_(bufsz)  CharT*       const __restrict buffer,
+           _In_range_(1, RSIZE_MAX)      rsize_t      const            bufsz,
+           _In_z_ _Printf_format_string_ const CharT* const __restrict format,
+           _In_                          Args&&...                     args)
+{
+#if _DEBUG
+    assert((buffer != nullptr) && (format != nullptr));
+    assert((bufsz > 0) && (bufsz <= RSIZE_MAX));
+    // too difficult as a generic
+    if constexpr(std::is_same_v<CharT, char>)
+    {
+        assert(std::string{ format }.find("%n") == std::string::npos);
+        std::string test{ format };
+        const auto num_specs{ std::count(test.begin(), test.end(), '%') };
+        assert(num_specs == sizeof...(args));
+    }
+    else if constexpr(std::is_same_v<CharT, wchar_t>)
+    {
+        assert(std::wstring{ format }.find(L"%n") == std::wstring::npos);
+        std::wstring test{ format };
+        const auto num_specs{ std::count(test.begin(), test.end(), L'%') };
+        assert(num_specs == sizeof...(args));
+    }
```

Comment:
It's generally bad practice to have different behavior in release vs debug builds. if one gets modified and the other doesn't, it's hard to diagnose the difference. Since we don't handle exceptions very well, I recommend limiting the problems to the asserts. However, as these are catastrophic errors, the exceptions would be justified.

Reply:
For now I'll just leave the asserts